### ISSUE TITLE
Implement the Golang version of the existing CPP Wasm plugin: key-rate-limit 

### ISSUE THREAD

**Issue by WeixinX** [alibaba/higress#596]

**Comment by ZackSoul** [comment]
I want to have a try

**Comment by ZackSoul** [comment]
Hello, I noticed that the rank of this issue was adjusted to challenge, so I'd like to ask who the committer of this issue is, and I have some questions to ask.

**Comment by ZackSoul** [comment]
![question](https://github.com/alibaba/higress/assets/91828262/d8303b18-ba21-4fb8-a8c3-58018a661798)
How can I get this rule_id when I implement it in go language?

**Comment by ZackSoul** [comment]
And, I noticed that the rank of this issue was adjusted to CHALLENGE, so shouldn't there be another developer involved, hopefully I can get some guidance, thanks!

**Comment by johnlanni** [comment]
Yeah, the rank was adjusted to `level/challenge` , because the wasm-go sdk needs to be extended to support it first.
The areas that need to be expanded are as follows:
1. Support the hostcall function `onTick` which binding to the vm context
2. Design a mechanism that can identify different rules using the same "limit_key"(apply to different domains or ingresses)

**Comment by ZackSoul** [comment]
我正在编写key-rate-limit的e2e测试，但是我使用goroutine并发进行测试时发现由于http.MakeRequestAndExpectEventuallyConsistentResponse函数为了到达requiredConsecutiveSuccesses = 3的要求时会重复发送请求，这导致我并不能按照我规定的请求数判断预期结果导致测试失败，不知道我的描述是否清晰，期望得到一点帮助，谢谢。

**Comment by ZackSoul** [comment]
期待回复和帮助

**Comment by johnlanni** [comment]
@ZackSoul 可以修改e2e测试框架从而来满足你的测试需求

**Comment by kai2321** [comment]
Is there still a need to work on this issue? @johnlanni I'd like to give it a try if you’re no longer continuing with it. @ZackSoul

**Comment by ZackSoul** [comment]
我很抱歉没能完成这个issue，你可以继续如果社区对于这个issue还有需求的话

**Comment by CH3CHO** [comment]
I've assigned this to you. Thanks. @kai2321

**Comment by kai2321** [comment]
Thanks， I will try my best to finish it

**Comment by kai2321** [comment]
请求review（如果有时间的话） @CH3CHO @johnlanni

**Comment by CH3CHO** [comment]
> 请求review（如果有时间的话） [@CH3CHO](https://github.com/CH3CHO) [@johnlanni](https://github.com/johnlanni)

好的~
